src/Microsoft.ML.ImageAnalytics/ImageLoader.cs

```diff
@@ -201,11 +239,112 @@ protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, b
+                ValueGetter<VBuffer<byte>> del =
+                    (ref VBuffer<byte> dst) =>
+                    {
+                        getSrc(ref src);
+
+                        if (src.Length > 0)
+                        {
+                            string path = src.ToString();
+                            if (!string.IsNullOrWhiteSpace(_parent.ImageFolder))
+                                path = Path.Combine(_parent.ImageFolder, path);
+
+                            int imgSize = LoadDataIntoBuffer(path, ref dst);
+                            if (imgSize < 0)
+                                throw Host.Except($"Failed to load image {src.ToString()}.");
+                        }
+                    };
+
                 return del;
             }
 
+            public static int LoadDataIntoBuffer(string path, ref VBuffer<byte> imgData)
+            {
+                int count = -1;
+                int bytesread = -1;
+                // bufferSize == 1 used to avoid unnecessary buffer in FileStream
+                using (FileStream fs = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 1))
+                {
+                    long fileLength = fs.Length;
+                    if (fileLength > int.MaxValue)
+                        throw new IOException($"File {path} too big to open.");
+                    else if (fileLength == 0)
+                    {
+                        Console.WriteLine("File length is zero even though the image is not empty");
+                        byte[] imageBuffer;
+                        // Some file systems (e.g. procfs on Linux) return 0 for length even when there's content.
+                        // Thus we need to assume 0 doesn't mean empty.
+                        imageBuffer = File.ReadAllBytes(path);
+                        count = imageBuffer.Length;
+                        imgData = new VBuffer<byte>(count, imageBuffer);
+                        return count;
+                    }
+
+                    count = (int)fileLength;
+                    var editor = VBufferEditor.Create(ref imgData, count);
+
+#if NETSTANDARD2_0
+                    bytesread = ReadToEnd(fs, editor.Values);
+                    Contracts.Assert(count == bytesread);
+
+#else
+                    fs.Read(editor.Values);
+                    bytesread = editor.Values.Length;
+                    Contracts.Assert(count == bytesread);
+#endif
+                    imgData = editor.Commit();
+                    return count;
+
+                }
+
+            }
+
+            public static int ReadToEnd(System.IO.Stream stream, Span<byte> bufferspan)
+            {
+
+                int chunksize = 4096; // Most optimal size for buffer, friendly to CPU's L1 cache
+                byte[] readBuffer = new byte[chunksize];
+                int totalBytesRead = 0;
+                int bytesRead;
+                unsafe
```

Comment:
You shouldn't use unsafe code to copy from a `byte[]` to a `Span<byte>`. Instead you can use [Span.CopyTo[([https://docs.microsoft.com/en-us/dotnet/api/system.span-1.copyto](https://docs.microsoft.com/en-us/dotnet/api/system.span-1.copyto))

Reply:
We will need to benchmark to make sure which is faster. Please benchmark what Eric is suggesting before taking this change. 

---
In reply to: [330658232](https://github.com/dotnet/machinelearning/pull/4242#discussion_r330658232) [](ancestors = 330658232)

Reply:
Can I use Span.Copy to copy from a byte[] ? I am unable to do that.

Reply:
All arrays have a `.ToSpan()` method to get a span from an array.

Reply:
Changed in the recent commit.